src/DeviceInput/deviceInputSystem.ts

```diff
@@ -291,13 +307,42 @@ export class DeviceInputSystem implements IDisposable {
             if (kbKey) {
                 kbKey[evt.keyCode] = 0;
                 if (this.onInputChanged) {
-                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode]);
+                    this.onInputChanged(DeviceType.Keyboard, 0, evt.keyCode, 1, kbKey[evt.keyCode], evt);
+                }
+            }
+        });
+
+        let attachedFunction = () => {
+            this._elementToAttachTo.addEventListener("keydown", this._keyboardDownEvent);
+            this._elementToAttachTo.addEventListener("keyup", this._keyboardUpEvent);
+        };
+
+        this._onCanvasFocusObserver = this._engine.onCanvasFocusObservable.add(
+            (() => {
+                if (document.activeElement === this._elementToAttachTo) {
+                    attachedFunction();
+                }
+                return attachedFunction;
+            })()
+        );

```

Comment:
I'm really confused by this code. What is this trying to do?

Reply:
Talked offline